sdk/python/kfp/compiler/compiler_utils.py

```diff
@@ -427,14 +427,29 @@ def get_dependencies(
                 task2=task,
             )
 
-            # a task cannot depend on a task created in a for loop group since individual PipelineTask variables are reassigned after each loop iteration
+            # ParralelFor Check
+            for parent in task_name_to_parent_groups[upstream_task.name]:
+                parent = group_name_to_group.get(parent, None)
+                if isinstance(
+                        parent,
+                        tasks_group.ParallelFor) and task not in parent.tasks:
+                    raise RuntimeError(
+                        f'Task {task.name} cannot dependent on any task inside a ParralelFor'
+                    )
+
+            # Condition check
+            dependent_group = group_name_to_group.get(upstream_groups[0], None)
+            if isinstance(dependent_group, tasks_group.Condition):
+                raise RuntimeError(
+                    f'Task {task.name} cannot dependent on any task inside a Condition that is not a common ancestor of both tasks'
+                )
+
+            # ExitHandler check
             dependent_group = group_name_to_group.get(upstream_groups[0], None)
```

Comment:
I'm not very familiar with the functions/objects in this part of the codebase. What is the first element of `upstream_groups`?

Reply:
The first element of upstream group is the first ancestor(condition, parallelfor or exithandler) that is not common between both tasks. So if that ancestor is a condition or exithandler we should fail